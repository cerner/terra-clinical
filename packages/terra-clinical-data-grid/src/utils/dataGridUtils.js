
/**
 * The PAGED_CONTENT_OFFSET_BUFFER represents the vertical space (in px) remaining in the vertical overflow
 * that will trigger additonal content retrieval (if onRequestContent is provided).
 */
const PAGED_CONTENT_OFFSET_BUFFER = 100;

/**
 * The Object representation of the void column. This empty column is used
 * as negative space to allow for resizing of the last overflow column.
 */
const VOID_COLUMN = {
  id: 'DataGrid-voidColumn',
  width: 150,
};

/**
 * The Object representation of the row selection column.
 */
const ROW_SELECTION_COLUMN = {
  id: 'DataGrid-rowSelectionColumn',
  width: 50,
};

/**
 * Returns an Array of HTMLElements that are children of the given 'element' parameter and have
 * the data-accessible-data-grid-content attribute defined.
 * @param {HTMLElement} element The element within which to search for accessible content.
 */
const getAccessibleContents = (element) => {
  const accessibleArray = [];
  const accessibleContentNodes = element.querySelectorAll('[data-accessible-data-grid-content]');

  for (let i = 0, numberOfNodes = accessibleContentNodes.length; i < numberOfNodes; i += 1) {
    accessibleArray.push(accessibleContentNodes[i]);
  }

  return accessibleArray;
};

/**
 * Returns the column's specified width or the default width if not width is defined.
 * @param {Object} column Object adhering to the columnData shape.
 * @param {Number} defaultColumnWidth Number (in px) indicating the default width to be used if a column's width is otherwise undefined.
 */
const getWidthForColumn = (column, defaultColumnWidth) => (column && column.width) || defaultColumnWidth;

/**
 * Returns the combined width of every column provided.
 * @param {Array} columns Array of Objects adhering to the columnData shape.
 * @param {Number} defaultColumnWidth Number (in px) indicating the default width to be used if a column's width is otherwise undefined.
 */
const getTotalColumnWidth = (columns, defaultColumnWidth) => {
  if (!columns) {
    return 0;
  }

  return columns.reduce((totalWidth, column) => totalWidth + getWidthForColumn(column, defaultColumnWidth), 0);
};

/**
 * Returns the pinned columns from the given source, including the row selection column if necessary.
 * @param {Object} props Object conforming to DataGrid's prop types.
 */
const getPinnedColumns = (props) => {
  const { pinnedColumns, hasSelectableRows } = props;

  let updatedPinnedColumns = pinnedColumns;
  if (hasSelectableRows) {
    updatedPinnedColumns = [ROW_SELECTION_COLUMN].concat(updatedPinnedColumns);
  }

  return updatedPinnedColumns;
};

/**
 * Returns the overflow columns from the given source, including the void column used for column resizing.
 * @param {Object} props Object conforming to DataGrid's prop types.
 */
const getOverflowColumns = (props) => {
  const { overflowColumns, hasResizableColumns } = props;

  if (hasResizableColumns) {
    return overflowColumns.concat([VOID_COLUMN]);
  }

  return overflowColumns;
};

/**
 * Returns true if the given element matches the given selector. Includes support for IE10.
 * @param {Element} element The element to compare against the selector.
 * @param {String} selector The selector string to test.
 */
const matchesSelector = (element, selector) => {
  if (Element.prototype.msMatchesSelector) {
    return element.msMatchesSelector(selector);
  }

  return element.matches(selector);
};

/**
 * Generates identifiers for accessible elements within the DataGrid.
 * @param {Object} props The DataGrid props.
 * @param {Object} headerCellRefs The references to accessible header cell elements.
 * @param {Object} sectionRefs The references to accessible section header elements.
 * @param {Object} cellRefs The references to accessible cell elements.
 */
const generateAccessibleContentIndex = (props, headerCellRefs, sectionRefs, cellRefs) => {
  const { sections } = props;

  const pinnedColumns = getPinnedColumns(props);
  const overflowColumns = getOverflowColumns(props);

  const orderedColumnIds = pinnedColumns.concat(overflowColumns).map(column => column.id);

  let accessibilityStack = [];

  pinnedColumns.forEach((column) => {
    const headerRef = headerCellRefs[column.id];
    if (headerRef) {
      if (column.isSelectable) {
        accessibilityStack.push(headerRef);
      }

      accessibilityStack = accessibilityStack.concat(getAccessibleContents(headerRef.parentNode));
    }
  });

  overflowColumns.forEach((column) => {
    const headerRef = headerCellRefs[column.id];

    if (headerRef) {
      if (column.isSelectable) {
        accessibilityStack.push(headerRef);
      }

      accessibilityStack = accessibilityStack.concat(getAccessibleContents(headerRef.parentNode));
    }
  });

  sections.forEach((section) => {
    const sectionRef = sectionRefs[section.id];

    if (sectionRef) {
      if (section.isCollapsible) {
        accessibilityStack.push(sectionRef);
      }

      accessibilityStack = accessibilityStack.concat(getAccessibleContents(sectionRef.parentNode));
    }

    if (section.isCollapsed) {
      /**
       * If the section is collapsed, we do not want to assign accessibility identifiers to its content.
       */
      return;
    }

    section.rows.forEach((row) => {
      const cellMap = {};
      row.cells.forEach((cell) => {
        cellMap[cell.columnId] = cell;
      });

      orderedColumnIds.forEach((columnId) => {
        const cellRef = cellRefs[`${section.id}-${row.id}-${columnId}`];
        if (cellRef) {
          if ((cellMap[columnId] && cellMap[columnId].isSelectable) || (columnId === 'DataGrid-rowSelectionColumn' && row.isSelectable)) {
            accessibilityStack.push(cellRef);
          }

          accessibilityStack = accessibilityStack.concat(getAccessibleContents(cellRef.parentNode));
        }
      });
    });
  });

  accessibilityStack.forEach((element, index) => {
    element.setAttribute('data-accessibility-id', index);
  });

  return accessibilityStack;
};

const dataGridUtils = {
  PAGED_CONTENT_OFFSET_BUFFER,
  VOID_COLUMN,
  ROW_SELECTION_COLUMN,
  getAccessibleContents,
  getWidthForColumn,
  getTotalColumnWidth,
  getPinnedColumns,
  getOverflowColumns,
  matchesSelector,
  generateAccessibleContentIndex,
};

export default dataGridUtils;
