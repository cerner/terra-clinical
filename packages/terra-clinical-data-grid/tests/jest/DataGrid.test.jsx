import React from 'react';

import DataGrid from '../../src/DataGrid';
import dataGridUtils from '../../src/utils/dataGridUtils';
import messages from '../../translations/en-US.json';

const testColumns = {
  'Column-0': {
    id: 'Column-0',
    width: 200,
    text: 'Column 0',
  },
  'Column-1': {
    id: 'Column-1',
    width: 200,
    text: 'Column 1',
  },
  'Column-2': {
    id: 'Column-2',
    width: 200,
    text: 'Column 2',
  },
  'Column-3': {
    id: 'Column-3',
    width: 200,
    text: 'Column 3',
  },
};

const testSections = [{
  id: 'section-0',
  rows: [{
    id: 'row-0',
    ariaLabel: 'Row 0',
    cells: [{
      columnId: 'Column-0',
      component: <div>0</div>,
    }, {
      columnId: 'Column-1',
      component: <div>1</div>,
    }, {
      columnId: 'Column-2',
      component: <div>2</div>,
    }, {
      columnId: 'Column-3',
      component: <div>3</div>,
    }],
  }, {
    id: 'row-1',
    ariaLabel: 'Row 1',
    cells: [{
      columnId: 'Column-0',
      component: <div>0</div>,
    }, {
      columnId: 'Column-1',
      component: <div>1</div>,
    }, {
      columnId: 'Column-2',
      component: <div>2</div>,
    }, {
      columnId: 'Column-3',
      component: <div>3</div>,
    }],
  }],
}];

const mockIntl = {
  formatDate: () => {},
  formatTime: () => {},
  formatRelative: () => {},
  formatNumber: () => {},
  formatPlural: () => {},
  formatMessage: ({ id }, values) => `${messages[id]}-${JSON.stringify(values)}`,
  formatHTMLMessage: () => {},
  now: () => {},
};

describe('DataGrid Snapshots', () => {
  it('should render a DataGrid with missing optional props', () => {
    const dataGrid = shallow(<DataGrid.WrappedComponent id="test" intl={mockIntl} />);
    expect(dataGrid).toMatchSnapshot();
  });

  it('should render a DataGrid with only overflow columns', () => {
    const dataGridComp = (
      <DataGrid.WrappedComponent
        id="test"
        overflowColumns={[testColumns['Column-0'], testColumns['Column-1'], testColumns['Column-2'], testColumns['Column-3']]}
        sections={testSections}
        fill
        intl={mockIntl}
      />
    );

    const dataGrid = shallow(dataGridComp);
    expect(dataGrid).toMatchSnapshot();
  });

  it('should render a DataGrid with pinned and overflow columns', () => {
    const dataGridComp = (
      <DataGrid.WrappedComponent
        id="test"
        pinnedColumns={[testColumns['Column-0'], testColumns['Column-1']]}
        overflowColumns={[testColumns['Column-2'], testColumns['Column-3']]}
        sections={testSections}
        fill
        intl={mockIntl}
      />
    );

    const dataGrid = shallow(dataGridComp);
    expect(dataGrid).toMatchSnapshot();
  });

  it('should render a DataGrid with custom row and header heights', () => {
    const dataGridComp = (
      <DataGrid.WrappedComponent
        id="test"
        overflowColumns={[testColumns['Column-0'], testColumns['Column-1'], testColumns['Column-2'], testColumns['Column-3']]}
        sections={testSections}
        rowHeight="5rem"
        headerHeight="10rem"
        fill
        intl={mockIntl}
      />
    );

    const dataGrid = shallow(dataGridComp);
    expect(dataGrid).toMatchSnapshot();
  });

  it('should render a DataGrid with selectable rows, columns, and cells', () => {
    const dataGridComp = (
      <DataGrid.WrappedComponent
        id="test"
        pinnedColumns={[{
          id: 'Column-0',
          width: 200,
          text: 'Column 0',
          isSelectable: true,
          isResizable: true,
          sortIndicator: 'ascending',
        }]}
        overflowColumns={[{
          id: 'Column-1',
          width: 300,
          text: 'Column 1',
          isSelectable: true,
          sortIndicator: 'descending',
        }]}
        sections={[{
          id: 'section-0',
          isCollapsible: true,
          isCollapsed: false,
          text: 'Section 0',
          startAccessory: <div>Start Accessory</div>,
          endAccessory: <div>End Accessory</div>,
          rows: [{
            id: 'row-0',
            isSelectable: true,
            isSelected: true,
            ariaLabel: 'Row 0',
            cells: [{
              columnId: 'Column-0',
              isSelectable: true,
              isSelected: true,
              component: <div>0</div>,
            }, {
              columnId: 'Column-1',
              isSelectable: true,
              component: <div>1</div>,
            }],
          }],
        }]}
        hasSelectableRows
        onColumnSelect={() => {}}
        onCellSelect={() => {}}
        onRowSelect={() => {}}
        fill
        intl={mockIntl}
      />
    );

    const dataGrid = shallow(dataGridComp);
    expect(dataGrid).toMatchSnapshot();
  });

  it('should render a DataGrid with custom header cells', () => {
    const dataGridComp = (
      <DataGrid.WrappedComponent
        id="test"
        pinnedColumns={[{
          id: 'Column-0',
          width: 200,
          text: 'Column 0',
          sortIndicator: 'ascending',
          component: <div>Custom Header 0</div>,
        }]}
        overflowColumns={[{
          id: 'Column-1',
          width: 300,
          text: 'Column 1',
          sortIndicator: 'descending',
          component: <div>Custom Header 1</div>,
        }]}
        sections={[{
          id: 'section-0',
          isCollapsible: true,
          isCollapsed: false,
          text: 'Section 0',
          startAccessory: <div>Start Accessory</div>,
          endAccessory: <div>End Accessory</div>,
          rows: [{
            id: 'row-0',
            cells: [{
              columnId: 'Column-0',
              component: <div>0</div>,
            }, {
              columnId: 'Column-1',
              component: <div>1</div>,
            }],
          }],
        }]}
        fill
        intl={mockIntl}
      />
    );

    const dataGrid = shallow(dataGridComp);
    expect(dataGrid).toMatchSnapshot();
  });

  it('should render a DataGrid with custom section header', () => {
    const dataGridComp = (
      <DataGrid.WrappedComponent
        id="test"
        pinnedColumns={[{
          id: 'Column-0',
          width: 200,
          text: 'Column 0',
          sortIndicator: 'ascending',
        }]}
        overflowColumns={[{
          id: 'Column-1',
          width: 300,
          text: 'Column 1',
          sortIndicator: 'descending',
        }]}
        sections={[{
          id: 'section-0',
          isCollapsible: true,
          isCollapsed: false,
          text: 'Section 0',
          startAccessory: <div>Start Accessory</div>,
          endAccessory: <div>End Accessory</div>,
          component: <div>Custom Section Header</div>,
          rows: [{
            id: 'row-0',
            cells: [{
              columnId: 'Column-0',
              component: <div>0</div>,
            }, {
              columnId: 'Column-1',
              component: <div>1</div>,
            }],
          }],
        }]}
        fill
        intl={mockIntl}
      />
    );

    const dataGrid = shallow(dataGridComp);
    expect(dataGrid).toMatchSnapshot();
  });
});

it('should render a DataGrid with the fill prop missing', () => {
  const dataGridComp = (
    <DataGrid.WrappedComponent
      id="test"
      pinnedColumns={[testColumns['Column-0'], testColumns['Column-1']]}
      overflowColumns={[testColumns['Column-2'], testColumns['Column-3']]}
      sections={testSections}
      intl={mockIntl}
    />
  );

  const dataGrid = shallow(dataGridComp);
  expect(dataGrid).toMatchSnapshot();
});

describe('getDerivedStateFromProps', () => {
  let getPinnedColumnsRef;
  let getOverflowColumnsRef;
  let getTotalColumnWidthRef;

  beforeEach(() => {
    getPinnedColumnsRef = dataGridUtils.getPinnedColumns;
    getOverflowColumnsRef = dataGridUtils.getOverflowColumns;
    getTotalColumnWidthRef = dataGridUtils.getTotalColumnWidth;
  });

  afterEach(() => {
    dataGridUtils.getPinnedColumns = getPinnedColumnsRef;
    dataGridUtils.getOverflowColumns = getOverflowColumnsRef;
    dataGridUtils.getTotalColumnWidth = getTotalColumnWidthRef;
  });

  it('should return the calculated pinned/overflow column widths', () => {
    const pinnedColumns = {};
    const overflowColumns = {};

    dataGridUtils.getPinnedColumns = jest.fn().mockReturnValue(pinnedColumns);
    dataGridUtils.getOverflowColumns = jest.fn().mockReturnValue(overflowColumns);
    dataGridUtils.getTotalColumnWidth = jest.fn().mockImplementation((columns) => {
      if (columns === pinnedColumns) {
        return 555;
      }

      if (columns === overflowColumns) {
        return 777;
      }
      return -1;
    });

    const result = DataGrid.WrappedComponent.getDerivedStateFromProps({}, 300);
    expect(result.pinnedColumnWidth).toEqual(555);
    expect(result.overflowColumnWidth).toEqual(777);
  });
});
