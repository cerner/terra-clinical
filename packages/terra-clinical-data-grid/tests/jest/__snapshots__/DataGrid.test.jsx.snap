// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGrid Snapshots should render a DataGrid with custom header cells 1`] = `
<InjectIntl(DataGrid)
  fill={true}
  id="test"
  overflowColumns={
    Array [
      Object {
        "component": <div>
          Custom Header 1
        </div>,
        "id": "Column-1",
        "sortIndicator": "descending",
        "text": "Column 1",
        "width": 300,
      },
    ]
  }
  pinnedColumns={
    Array [
      Object {
        "component": <div>
          Custom Header 0
        </div>,
        "id": "Column-0",
        "sortIndicator": "ascending",
        "text": "Column 0",
        "width": 200,
      },
    ]
  }
  sections={
    Array [
      Object {
        "endAccessory": <div>
          End Accessory
        </div>,
        "id": "section-0",
        "isCollapsed": false,
        "isCollapsible": true,
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
            ],
            "id": "row-0",
          },
        ],
        "startAccessory": <div>
          Start Accessory
        </div>,
        "text": "Section 0",
      },
    ]
  }
/>
`;

exports[`DataGrid Snapshots should render a DataGrid with custom row and header heights 1`] = `
<InjectIntl(DataGrid)
  fill={true}
  headerHeight="10rem"
  id="test"
  overflowColumns={
    Array [
      Object {
        "id": "Column-0",
        "text": "Column 0",
        "width": 200,
      },
      Object {
        "id": "Column-1",
        "text": "Column 1",
        "width": 200,
      },
      Object {
        "id": "Column-2",
        "text": "Column 2",
        "width": 200,
      },
      Object {
        "id": "Column-3",
        "text": "Column 3",
        "width": 200,
      },
    ]
  }
  rowHeight="5rem"
  sections={
    Array [
      Object {
        "id": "section-0",
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
              Object {
                "columnId": "Column-2",
                "component": <div>
                  2
                </div>,
              },
              Object {
                "columnId": "Column-3",
                "component": <div>
                  3
                </div>,
              },
            ],
            "id": "row-0",
          },
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
              Object {
                "columnId": "Column-2",
                "component": <div>
                  2
                </div>,
              },
              Object {
                "columnId": "Column-3",
                "component": <div>
                  3
                </div>,
              },
            ],
            "id": "row-1",
          },
        ],
      },
    ]
  }
/>
`;

exports[`DataGrid Snapshots should render a DataGrid with custom section header 1`] = `
<InjectIntl(DataGrid)
  fill={true}
  id="test"
  overflowColumns={
    Array [
      Object {
        "id": "Column-1",
        "sortIndicator": "descending",
        "text": "Column 1",
        "width": 300,
      },
    ]
  }
  pinnedColumns={
    Array [
      Object {
        "id": "Column-0",
        "sortIndicator": "ascending",
        "text": "Column 0",
        "width": 200,
      },
    ]
  }
  sections={
    Array [
      Object {
        "component": <div>
          Custom Section Header
        </div>,
        "endAccessory": <div>
          End Accessory
        </div>,
        "id": "section-0",
        "isCollapsed": false,
        "isCollapsible": true,
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
            ],
            "id": "row-0",
          },
        ],
        "startAccessory": <div>
          Start Accessory
        </div>,
        "text": "Section 0",
      },
    ]
  }
/>
`;

exports[`DataGrid Snapshots should render a DataGrid with missing optional props 1`] = `
<InjectIntl(DataGrid)
  id="test"
/>
`;

exports[`DataGrid Snapshots should render a DataGrid with only overflow columns 1`] = `
<InjectIntl(DataGrid)
  fill={true}
  id="test"
  overflowColumns={
    Array [
      Object {
        "id": "Column-0",
        "text": "Column 0",
        "width": 200,
      },
      Object {
        "id": "Column-1",
        "text": "Column 1",
        "width": 200,
      },
      Object {
        "id": "Column-2",
        "text": "Column 2",
        "width": 200,
      },
      Object {
        "id": "Column-3",
        "text": "Column 3",
        "width": 200,
      },
    ]
  }
  sections={
    Array [
      Object {
        "id": "section-0",
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
              Object {
                "columnId": "Column-2",
                "component": <div>
                  2
                </div>,
              },
              Object {
                "columnId": "Column-3",
                "component": <div>
                  3
                </div>,
              },
            ],
            "id": "row-0",
          },
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
              Object {
                "columnId": "Column-2",
                "component": <div>
                  2
                </div>,
              },
              Object {
                "columnId": "Column-3",
                "component": <div>
                  3
                </div>,
              },
            ],
            "id": "row-1",
          },
        ],
      },
    ]
  }
/>
`;

exports[`DataGrid Snapshots should render a DataGrid with pinned and overflow columns 1`] = `
<InjectIntl(DataGrid)
  fill={true}
  id="test"
  overflowColumns={
    Array [
      Object {
        "id": "Column-2",
        "text": "Column 2",
        "width": 200,
      },
      Object {
        "id": "Column-3",
        "text": "Column 3",
        "width": 200,
      },
    ]
  }
  pinnedColumns={
    Array [
      Object {
        "id": "Column-0",
        "text": "Column 0",
        "width": 200,
      },
      Object {
        "id": "Column-1",
        "text": "Column 1",
        "width": 200,
      },
    ]
  }
  sections={
    Array [
      Object {
        "id": "section-0",
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
              Object {
                "columnId": "Column-2",
                "component": <div>
                  2
                </div>,
              },
              Object {
                "columnId": "Column-3",
                "component": <div>
                  3
                </div>,
              },
            ],
            "id": "row-0",
          },
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
              },
              Object {
                "columnId": "Column-2",
                "component": <div>
                  2
                </div>,
              },
              Object {
                "columnId": "Column-3",
                "component": <div>
                  3
                </div>,
              },
            ],
            "id": "row-1",
          },
        ],
      },
    ]
  }
/>
`;

exports[`DataGrid Snapshots should render a DataGrid with selectable rows, columns, and cells 1`] = `
<InjectIntl(DataGrid)
  fill={true}
  hasSelectableRows={true}
  id="test"
  onCellSelect={[Function]}
  onColumnSelect={[Function]}
  onRowSelect={[Function]}
  overflowColumns={
    Array [
      Object {
        "id": "Column-1",
        "isSelectable": true,
        "sortIndicator": "descending",
        "text": "Column 1",
        "width": 300,
      },
    ]
  }
  pinnedColumns={
    Array [
      Object {
        "id": "Column-0",
        "isResizable": true,
        "isSelectable": true,
        "sortIndicator": "ascending",
        "text": "Column 0",
        "width": 200,
      },
    ]
  }
  sections={
    Array [
      Object {
        "endAccessory": <div>
          End Accessory
        </div>,
        "id": "section-0",
        "isCollapsed": false,
        "isCollapsible": true,
        "rows": Array [
          Object {
            "cells": Array [
              Object {
                "columnId": "Column-0",
                "component": <div>
                  0
                </div>,
                "isSelectable": true,
                "isSelected": true,
              },
              Object {
                "columnId": "Column-1",
                "component": <div>
                  1
                </div>,
                "isSelectable": true,
              },
            ],
            "id": "row-0",
            "isSelectable": true,
            "isSelected": true,
          },
        ],
        "startAccessory": <div>
          Start Accessory
        </div>,
        "text": "Section 0",
      },
    ]
  }
/>
`;
